### ISSUE TITLE
[HOLD for payment 2024-04-25] [Tracking] [Performance] Enable the new Fabric architecture

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#8503]
GitHub Project: https://github.com/orgs/Expensify/projects/35

Main New Arch PR: https://github.com/Expensify/App/pull/13767

## Context 
New render engine: https://reactnative.dev/architecture/fabric-renderer
Summary of findings 
1) There are 4 main components of RN: React, JavaScript, a series of elements collectively known as the “bridge”, and the native elements/modules. Each of these components are getting complimentary upgrades.
2) The notable React upgrades discussed are concurrent rendering, synchronous event callbacks, and priority rendering queues. Together those improvements basically allow for the UI thread to perform heavy rendering work in the background without compromising quick response times to user events such as scrolling and navigation gestures.
3) In the existing RN architecture, the JS and native sides are essentially unaware of each other. All communications are serialized JSON messages sent asynchronously over the bridge.
4) When JavaScript communicates with a web browser, the browser objects hold direct references to native input elements which are controlled internally by the browser. So an HTML checkbox in a browser actually internally holds a reference to a native macOS or Windows checkbox that’s managed by the browser. Similarly, JSI (JavaScript Interface) is a new piece of RN that enables direct communication between JS and the native side, because JSI internally manages direct references to C++ host objects (similar to the HTMLInput element on a browser). This means that messages sent between JS and native modules no longer need to be JSON-serialized or sent over the single bottleneck communication channel. It also means messages can be sent either synchronously or asynchronously.
6) At a high level, the traditional RN bridge manages two tasks: computing layout via a component tree called the “shadow tree” (that shadows the React tree), and managing native modules such as the camera. In the new architecture, these are broken into two distinct pieces which facebook is calling Fabric and TurboModules, respectively.
6) Fabric, the new UI manager, leverages JSI to create the shadow tree in C++ directly, cutting out the middle man and dramatically improving performance. It also exposes UI operations on both sides, so the native side and JS side can both directly manipulate the shadow tree without having to send serialized messages back and forth across the bridge.
7) In the existing native module system, all the native modules need to be loaded when the app launches so that they’re ready and waiting for commands to be sent over the bridge. Turbo Modules have direct references to native modules via JSI, which gives two main benefits. First, native modules can be lazy-loaded (only when really needed), which should dramatically improve app startup time, especially on apps that use lots of different native modules. Second, once loaded the modules will be more performant because communications no longer need to be JSON-serialized and sent across the bridge.
8) Lastly, there is a tool I don’t fully understand called CodeGen. It leverages TypeScript to generate the interface files needed by Fabric and TurboModules. This compile-time type safety means that the code counterparts from both the native and JS worlds can trust eachother without any runtime checks, which means smaller bundle sizes and faster execution.

### Migrate off setNativeProps
- [x] https://github.com/Expensify/App/issues/11049
- [x] https://github.com/Expensify/App/issues/11050
- [x] https://github.com/Expensify/App/issues/11052
- [x] https://github.com/Expensify/App/issues/11053
- [x] https://github.com/Expensify/App/issues/11054
- [x] https://github.com/Expensify/App/issues/11055
- [x] https://github.com/Expensify/App/issues/11056

### Migrate off findNodeHandle
- [x] https://github.com/Expensify/App/issues/11635

### Bidirectional scrolling 
- [x] testing and making sure the bidirectional pagination features we’ve built work with Fabric https://github.com/Expensify/react-native/pull/32

### Dependency upgrades
- [x] Upgrade to Reanimated 3 https://github.com/Expensify/App/pull/11567
- [x] react-native-svg https://github.com/Expensify/App/pull/12140
- [x] Upgrade react-native https://github.com/Expensify/App/issues/12177
- [x] [React Native Picker](https://github.com/react-native-picker/picker)
    - Has PR: https://github.com/react-native-picker/picker/pull/456
- [x] [progress-bar-android](https://github.com/react-native-progress-view/progress-bar-android)
    - Has PR: https://github.com/react-native-progress-view/progress-bar-android/pull/90/
- [x] [React Native PDF](https://github.com/wonday/react-native-pdf)
    - Has PR: https://github.com/wonday/react-native-pdf/pull/700
- [x] [UrbanAirship](https://github.com/urbanairship/react-native-module)
- [ ] ~~[FastImage](https://github.com/DylanVann/react-native-fast-image)~~ no longer in use
    - Has PR: https://github.com/DylanVann/react-native-fast-image/pull/951

### Known bugs
- [ ] **Bold fonts** don't work with Fabric enabled
- [x] https://github.com/Expensify/App/issues/14365
- [x] https://github.com/Expensify/App/issues/14366
- [x] https://github.com/Expensify/App/issues/14367
- [x] https://github.com/Expensify/App/issues/14368
- [x] https://github.com/Expensify/App/issues/14369
- [x] https://github.com/Expensify/App/issues/14370
- [x] https://github.com/Expensify/App/issues/14371
- [x] https://github.com/Expensify/App/issues/14372
- [x] https://github.com/Expensify/App/issues/16795
- [x] https://github.com/Expensify/App/issues/16796
- [x] https://github.com/Expensify/App/issues/16799
- [x] https://github.com/Expensify/App/issues/16806
- [x] https://github.com/Expensify/App/issues/16807
- [x] https://github.com/Expensify/App/issues/16808
- [x] https://github.com/Expensify/App/issues/16809
- [x] https://github.com/Expensify/App/issues/16810
- [x] https://github.com/Expensify/App/issues/16811
- [x] https://github.com/Expensify/App/issues/16814
- [x] https://github.com/Expensify/App/issues/16815
- [x] https://github.com/Expensify/App/issues/16824

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @roryabraham 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1648743262264529

**Comment by roryabraham** [comment]
This can definitely be external. I went ahead and flagged it as a potential issue for the Margelo team to work on: https://expensify.slack.com/archives/C035J5C9FAP/p1649217447621559

**Comment by mrousavy** [comment]
Awesome, Margelo would love to work on this! We noticed that [you use a lot of third party React Native libraries](https://github.com/Expensify/App/blob/1b0687408896efd9a1e0808d2f7b08b518846760/package.json#L79-L103), so it'll be a bit difficult to get those migrated as well. Especially problematic is Reanimated, but the SWM team is working on bringing that to Fabric.

Let's coordinate internally which issue we want to prioritize :)

**Comment by stitesExpensify** [comment]
Making this a weekly while we prioritize

**Comment by stitesExpensify** [comment]
Moving to monthly

**Comment by stitesExpensify** [comment]
@mrousavy just for clarity on this issue, where is it currently prioritized?

**Comment by mrousavy** [comment]
I believe currently we are not working on this issue, maybe let's discuss this in the Slack channel? We are currently working a lot with Coinbase and have very limited availability, but we expect to have some in two weeks!

**Comment by kidroca** [comment]
I've stumbled on this ticket thanks to experimenting with Fabric and the new RN Architecture.
It turns out migrating would resolve one top paying issue here: https://github.com/Expensify/App/issues/8349#issuecomment-1151172897

I think we can
- update react native using the migration tool here: https://react-native-community.github.io/upgrade-helper/
- audit which 3rd party dependencies released versions that support fabric and which didn't
- Then gradually upgrade towards fabric support

I'm posting a comment to keep in touch 
I'll be happy to work or take part in landing the new architecture to App

**Comment by stitesExpensify** [comment]
Thanks for volunteering @kidroca! There is still a lot of internal conversation happening around this, so I'll post an update when I have one 😄

**Comment by roryabraham** [comment]
Relevant discussion: https://expensify.slack.com/archives/C035J5C9FAP/p1655407547697249

**Comment by stitesExpensify** [comment]
No updateyet

**Comment by stitesExpensify** [comment]
@roryabraham are we still planning on doing this?

**Comment by roryabraham** [comment]
yes! We are now on our forked equivalent of the upstream RN 0.69.3. As soon as we _can_ turn on Fabric without introducing regressions, we absolutely should!

**Comment by roryabraham** [comment]
Oh, nice! We should get ahead of that and migrate away from `setNativeProps` usages ASAP.

**Comment by stitesExpensify** [comment]
Agreed. @rushatgabhane i'll create an issue, do you want me to just assign you, or would you prefer a contributor take it on?

**Comment by roryabraham** [comment]
Another one:

> Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot

**Comment by roryabraham** [comment]
Will also need to redo https://github.com/Expensify/App/pull/10334 before we can turn on Fabric (it caused a crash as was reverted)

**Comment by trjExpensify** [comment]
@stitesExpensify is out of the office this week, FYI. Feel free to pull the trigger on puller bear for the internal assignment portion.

**Comment by trjExpensify** [comment]
Yeah, it might be helpful to start treating this issue more like a main/tracking issue. The OP containing a checklist of the prerequisite issues we need to complete to be in a position to actually enable Fabric. That said, outside of deprecating `SetNativeProps`, do we have any other prerequisites that need to be completed at this point?

**Comment by luacmartins** [comment]
I agree with @trjExpensify that we should use this issue as a tracking issue. I went ahead and created separate issues for each `setNativeProps` migration and assigned them to you @rushatgabhane. I also updated the OP to track them and the existing PRs to link to the issues as well.

@rushatgabhane are there any other existing issues or PRs for the other requirements in these lists? 

https://expensify.slack.com/archives/C01GTK53T8Q/p1662132372464699?thread_ts=1662114154.919809&cid=C01GTK53T8Q

https://expensify.slack.com/archives/C01GTK53T8Q/p1662162747229869?thread_ts=1662114154.919809&cid=C01GTK53T8Q

**Comment by marcaaron** [comment]
Quick question here and sorry if it's already been mentioned...

But why is `setNativeProps()` not supported in fabric? And rather than refactor everything to not use it - did we explore just stubbing it in somehow so we can unlock using fabric now and then change all the components later?

**Comment by marcaaron** [comment]
Seems like there is some related stuff [here](https://github.com/facebook/react-native/issues/34391) and [here](https://reactnative.dev/docs/new-architecture-library-intro#migrating-off-setnativeprops). Does fabric pretty much just solve all the performance issues that led people to use uncontrolled components in the first place?

**Comment by luacmartins** [comment]
>But why is setNativeProps() not supported in fabric?

Not too sure. The [docs](https://reactnative.dev/docs/new-architecture-library-intro#migrating-off-setnativeprops) just mention that it's not supported 😭 They also [mention](https://reactnative.dev/docs/direct-manipulation) that Fabric Native Components achieve similar results to setNativeProps.

>did we explore just stubbing it in somehow so we can unlock using fabric now and then change all the components later?

I didn't explore any alternative solutions yet, but maybe @rushatgabhane did? I remember that when implementing Form `setNativeProps` was the only solution that worked on native without using state, but I don't recall the specific reason.

**Comment by marcaaron** [comment]
It would be good to rule out any obvious alternatives, but nothing is really coming to mind. Not entirely sure how you'd implement a `setNativeProps()`. Maybe it is less work than I am imagining it will be.

Maybe the best thing is to just start with one migration and verify performance does not get worse. Then again now sure how easy that will be since we'd maybe need to refactor all components to even start using the new version 😅

**Comment by rushatgabhane** [comment]
Asked about in react-native-new-architecture working group. https://github.com/reactwg/react-native-new-architecture/discussions/77

**Comment by roryabraham** [comment]
Going to co-assign here and help out, particularly when it comes to upgrading our fork as needed. Working on a PR to upgrade to `0.70.2` (based on `0.70.1` in the upstream)

**Comment by trjExpensify** [comment]
This is tied to WN, so I'm updating the priority label on this tracker to `weekly` for updates. [As per this thread](https://expensify.slack.com/archives/C02HWMSMZEC/p1664287165915339), adding a bullet to the OP for: 

- [ ] testing and making sure the bidirectional pagination features we’ve built work with Fabric.

**Comment by roryabraham** [comment]
Added a new section for dependency upgrades we'll need. One that stands out is that we'll need to [upgrade to Reanimated 3](https://blog.swmansion.com/announcing-reanimated-3-16167428c5f7), which is still in beta but will probably enter LTS soon

> Reanimated 3 will be the first version of the library that supports the new React Native architecture — Fabric

**Comment by roryabraham** [comment]
Updated this issue to include a requirement to address this warning:

```
Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17
```

Unless I'm mistaken, we can't get the benefits from Fabric unless we are actually using React 18

**Comment by kidroca** [comment]
> Updated this issue to include a requirement to address this warning:
> 
> ```
> Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17
> ```
> 
> Unless I'm mistaken, we can't get the benefits from Fabric unless we are actually using React 18

This is a `react-dom` (renderer) warning. Fabric is only for `react-native` (renderer)
We should still address it, but it's not relevant to the Fabric architecture
_(React 18 is the core, but the renderer implementations like `native` and `dom` are independent)_

**Comment by roryabraham** [comment]
Added https://github.com/Expensify/App/issues/11635 to the issue description

**Comment by roryabraham** [comment]
I'm going to create a separate tracking issue to enable React 18 on all platforms. Completing this issue and turning on Fabric will be one of those requirements.

**Comment by stitesExpensify** [comment]
Almost done migrating off of setNativeProps. Other requirements are also moving along!

**Comment by luacmartins** [comment]
https://github.com/Expensify/App/pull/11673 was merged, but the issue is still open for some reason. I think we can just close it.

**Comment by stitesExpensify** [comment]
Cool! Thanks guys!

**Comment by roryabraham** [comment]
The last two items are moving along nicely

- @janicduplessis is making progress on upgrading bidirectional scrolling to work with Fabric and the [PR](https://github.com/Expensify/react-native/pull/32) is in review.
- it sounds like @hannojg figured out how to finish the Reanimated 3 upgrade without us waiting for the official release. So that should be almost ready to move forward as well

Then comes the big step – actually make the switch to Fabric and test like we've never tested before to make sure we don't break stuff. I'd recommend we get like 3 C+ reviewers on that PR to make sure we have lots of eyes on it.

**Comment by trjExpensify** [comment]
Niceee! 🚀 Let's link those in the OP - I've added them both.

>I'd recommend we get like 3 C+ reviewers on that PR to make sure we have lots of eyes on it.

Yeah, I agree with this sentiment.

**Comment by JmillsExpensify** [comment]
Totally, the more C+ the merrier to make sure this is smooth.

**Comment by roryabraham** [comment]
> testing and making sure the bidirectional pagination features we’ve built work with Fabric

This was done in [this PR](https://github.com/Expensify/react-native/pull/32) and our bidirectional pagination features have been upgraded to work with Fabric. All that's left to check this item off is upgrading `E/App` to version `0.70.5` of our fork.

**Comment by hannojg** [comment]
Fixed the tests on the REA 3 PR, should be good now:
https://github.com/Expensify/App/pull/11567

**Comment by JmillsExpensify** [comment]
Still working through REA 3 PR.

**Comment by JmillsExpensify** [comment]
Expecting to [merge the PR](https://github.com/Expensify/App/pull/11567#issuecomment-1290028472) today.

**Comment by roryabraham** [comment]
REA 3 upgrade was merged yesterday 🎉 

But we discovered another blocker – react-native-svg does not yet support TurboModules, [WIP implementation here](https://github.com/software-mansion/react-native-svg/pull/1847)

**Comment by AndrewGable** [comment]
On iOS with Fabric + Turbo Modules I am experiencing an issue while navigating backwards from a chat to the Left Hand Nav. Then once I try to click back into the chat, the Left Hand Nav "flashes" as well. 

https://user-images.githubusercontent.com/2838819/198100114-e5be5491-ecac-44e2-95ec-874e6a841c0b.mov

**Comment by AndrewGable** [comment]
@janicduplessis Fixed the Android build issues, I opened a WIP PR linked above with those fixes included. The "flickering" issue occurs on both Android & iOS now.

**Comment by roryabraham** [comment]
REA 3 upgrade is complete

**Comment by roryabraham** [comment]
Looks like we need to revert https://github.com/Expensify/App/pull/12140 because it's causing a crash. So we'll need to work that crash out and merge it back in order to get Fabric enabled

**Comment by roryabraham** [comment]
According to [this thread](https://expensify.slack.com/archives/C04878MDF34/p1666864046673779), it sounds like we need to upgrade our react-native fork

**Comment by roryabraham** [comment]
We should investigate if react-native-pdf needs to be upgraded too

**Comment by JmillsExpensify** [comment]
@roryabraham What's blocking us at this point now that https://github.com/Expensify/App/issues/12177 is merged?

**Comment by JmillsExpensify** [comment]
Removing this from WAQ in light of [this discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1667276974060859). We'll continue to work on this outside of the immediate product roadmap though, for clarity.

**Comment by roryabraham** [comment]
Posted [here](https://expensify.slack.com/archives/C04878MDF34/p1667955504136409) to propose handing this issue off

**Comment by j-piasecki** [comment]
Hi 👋, I'm Jakub from SWM. Together with @WoLewicki, we'll be working on enabling the new architecture.

**Comment by WoLewicki** [comment]
Hi, I'm Wojciech from SWM, and I'll be working with @j-piasecki 😅

**Comment by AndrewGable** [comment]
Welcome welcome! Excited to see the results.

**Comment by roryabraham** [comment]
Cross-posting updates from slack – sounds like the main thing happening this week is [working on a fabric implementation in react-native-picker/picker](https://github.com/react-native-picker/picker/pull/456)

**Comment by roryabraham** [comment]
Regular updates coming in on slack, but it sounds like the main things happening right now are:

- Migrating picker
- Migrating `progress-bar-android`

**Comment by roryabraham** [comment]
Going to unassign @stitesExpensify and myself from this, because this is being handled by @j-piasecki and @WoLewicki going forward.

**Comment by WoLewicki** [comment]
Libraries with native components to migrate (with corresponding PRs if present already):
- [`react-native-picker/picker`](https://github.com/react-native-picker/picker/pull/456) ✅ 
- [`react-native-progress-view/progress-bar-android`](https://github.com/react-native-progress-view/progress-bar-android/pull/90/) ✅ 
- [`wonday/react-native-pdf`](https://github.com/wonday/react-native-pdf/pull/700) ✅ 
- [`urbanairship/react-native-module`](https://github.com/urbanairship/react-native-module)TBD
- [`DylanVann/react-native-fast-image`](https://github.com/DylanVann/react-native-fast-image)TBD

**Comment by roryabraham** [comment]
Asked the folks at UrbanAirship if they have any plans of upgrading https://github.com/urbanairship/react-native-module any time soon via our shared slack channel [here](https://expensify.slack.com/archives/C01RJ1RCK8D/p1669744967294199)

**Comment by dylanexpensify** [comment]
Coming from [this](https://expensify.slack.com/archives/C035J5C9FAP/p1673978436887229?thread_ts=1673373976.683909&cid=C035J5C9FAP) convo, added issues to the `Known Bug` section

**Comment by roryabraham** [comment]
Going to re-assign myself to this issue so that I can work on upgrading urbanairship-react-native to the [new @ua/react-native-airship package](https://www.npmjs.com/package/@ua/react-native-airship), on the old architecture to start.

**Comment by Julesssss** [comment]
Any updates here, or is there another issue to be following?

**Comment by roryabraham** [comment]
There have been many incremental updates in the #expensify-swm channel. I think there are only a few other libraries left to migrate to TurboModules – react-native-firebase/(analytics/perf/crashlytics), then every library we currently use will be Fabric and TurboModules-enabled.

I just kicked off a [new test build](https://github.com/Expensify/App/actions/runs/4376529492) and then we'll have Applause run another full regression suite against the Fabric-enabled app.

**Comment by roryabraham** [comment]
Starting a new round of Fabric testing: https://expensify.slack.com/archives/C9YU7BX5M/p1680021248483969

**Comment by roryabraham** [comment]
I believe Applause will be posting their summary of the second round of regression tests in a few minutes. When they do, @j-piasecki @WoLewicki if you could please comment on each issue so that I can assign them to you, that would be great. Thanks! (divide them up however you want)

Also, I'm thinking that maybe we should start using a GitHub Project to track this project?

**Comment by mvtglobally** [comment]
Fabric testing is completed 

QA identified multiple issues 

**Android** 
Deploy blocker - https://github.com/Expensify/App/issues/16692 
Since user is not able to login, we were not able to validate additional functionality 

**iOS**
 
Issues found 
https://github.com/Expensify/App/issues/16795
https://github.com/Expensify/App/issues/16796
https://github.com/Expensify/App/issues/16799
https://github.com/Expensify/App/issues/16806
https://github.com/Expensify/App/issues/16807
https://github.com/Expensify/App/issues/16808
https://github.com/Expensify/App/issues/16809
https://github.com/Expensify/App/issues/16810
https://github.com/Expensify/App/issues/16811
https://github.com/Expensify/App/issues/16814
https://github.com/Expensify/App/issues/16815
https://github.com/Expensify/App/issues/16824

**Comment by WoLewicki** [comment]
@roryabraham we assigned ourselves to all issues. As for the Github Project, yeah, why not? 😅

**Comment by roryabraham** [comment]
Requested another round of testing on the Fabric branch [here](https://expensify.slack.com/archives/C9YU7BX5M/p1683668064247209)

**Comment by roryabraham** [comment]
@j-piasecki @WoLewicki I think this issue has been going on long enough that we should schedule a call to chat about it and get aligned on where we're at w/ it and how best to move forward.

It seems like at this point we've logged a ton of issues and hopefully fixed many of them, so the next step now should be:

1. Make sure the fabric branch is up-to-date and has all conflicts resolved.
1. Manually retest all the reported issues tracked in the [GitHub Project](https://github.com/orgs/Expensify/projects/35), make sure they've been addressed
1. Run a new test build
1. Have Applause do another round of QA on the Fabric-enabled app
1. Hopefully merge the PR

**Comment by roryabraham** [comment]
Created a [LettuceMeet poll](https://lettucemeet.com/l/p5erB) to schedule a sync for the second half of next week.

**Comment by roryabraham** [comment]
Copying the summary from our call:

- Currently only two of the known issues are reproducible and they're minor
- There is a tricky bug w/ Freeze, stemming from the way Reanimated interacts with the shadow tree. As I understand it, @tomekzaw has a solution in mind. Maybe he can provide more context there.
- We want to upgrade to 0.72 before putting the Fabric upgrade up for review because it includes bug fixes, config changes, and new architecture performance improvements
- The last thing we'll want to do before putting the New Arch PR up for review is a quick audit of all the patches that we have there to remove those that aren't needed, make sure we have upstream PRs for those that are, and to list the upstream PR for each patch in the PR.
- We feel that the branch is stable enough and the bugs are minor enough that we aren't going to do another round of Applause QA in a test build. Instead, we'll focus on getting this actually merged, (hopefully within the next few weeks? :fingers_crossed:)

**Comment by tomekzaw** [comment]
> There is a tricky bug w/ Freeze, stemming from the way Reanimated interacts with the shadow tree. As I understand it, @tomekzaw has a solution in mind. Maybe he can provide more context there.

@j-piasecki investigated the regression in the app and managed to prepare a minimal repro of the underlying issue with Reanimated 3.1.0. The problem seems to be gone if we use the new algorithm for updating props on Fabric (https://github.com/software-mansion/react-native-reanimated/pull/4075) which has just been merged and will be released in 3.4.0.

**Comment by roryabraham** [comment]
Ok awesome, so adding to the above list that we need to upgrade to Reanimated 3.4.0

**Comment by roryabraham** [comment]
This is currently blocked on upgrading to RN 0.72

**Comment by roryabraham** [comment]
@WoLewicki and @j-piasecki are working on [adding new arch support for the MapBox RN SDK](https://github.com/rnmapbox/maps/pull/3022)

**Comment by roryabraham** [comment]
Last discussion around the upgrade is [here](https://expensify.slack.com/archives/C05LX9D6E07/p1696465592925979)

**Comment by roryabraham** [comment]
Discussing the last major blocker in slack: https://expensify.slack.com/archives/C04878MDF34/p1699892650055549

**Comment by roryabraham** [comment]
Seems we are finally very close to being able to turn on the new arch. Two PRs need to be merged first, in this order:

1. https://github.com/Expensify/App/pull/30905
1. https://github.com/Expensify/App/pull/31558

Then we can follow-up with the main PR to enable the new arch. However, after that there are (at least) two more follow-ups we'll want to work on:

1. A PR to enable bridgeless mode, such that no vestiges of the old arch remain
1. Work on enabling React Concurrent Mode so we can start using React 18 concurrency features such as `startTransition` and suspense

**Comment by WoLewicki** [comment]
I think 2. can be already worked on (if we mean web) so it is ready when new arch is merged @roryabraham.

**Comment by roryabraham** [comment]
@WoLewicki and @j-piasecki are unblocked and making progress on integrating RN 0.73 into the Fabric branch. some recent slack updates:

- https://expensify.slack.com/archives/C04878MDF34/p1705595312997069
- https://expensify.slack.com/archives/C04878MDF34/p1705682748348929

**Comment by roryabraham** [comment]
Moving this to weekly. let's get the New Arch merged!!!

**Comment by roryabraham** [comment]
Testing and development still happening very actively: https://github.com/Expensify/App/pull/13767#issuecomment-1924470912

**Comment by roryabraham** [comment]
Put together a [massive summary of outstanding dependency upgrades and patches](https://expensify.slack.com/archives/C01GTK53T8Q/p1707293299589539), along with recommended next steps for each. Hopefully that will help us track this better. It was kind of an unconventional format/too big for slack, so I'm open to suggestions on how to track this better.

**Comment by roryabraham** [comment]
- We've got a number of WIP dependency upgrades being tracked in the [project](https://github.com/orgs/Expensify/projects/35/views/1), so that's good to see some progress getting over the finish line.
- @ishpaul777 completed a lot of testing and found no more issues on iOS, will continue testing on Android
- @j-piasecki put together a [big list](https://github.com/Expensify/App/pull/13767#issuecomment-1943445518) of all the known issues from the New Arch, past and present. We need to put in some work to filter through them, find out which are still applicable in today's app, then expand them inline with reproduction steps, expected and actual results. We can use nested details tags for this huge list of test steps. I know this will be a grind and not be fun, but I think it's necessary

**Comment by roryabraham** [comment]
The following libs have been upgraded:

- react-native-image-picker
- react-native-haptic-feedback
- react-native-picker-select
- react-native-document-picker
- react-native-safe-area-context
- @react-native-community/clipboard
- react-native-android-location-enabler
- @rnmapbox/maps

I think a few more still pending:

- Reanimated
- react-native-camera-roll (upgrade was reverted due to a regression)
- react-native-image-manipulator => expo-image-manipulator

**Comment by cubuspl42** [comment]
Do I understand correctly that switching to Fabric in Expensify is a matter of weeks / 1-2 months?

**Comment by roryabraham** [comment]
that seems plausible

**Comment by roryabraham** [comment]
few misc updates:

- We created https://github.com/Expensify/App/issues/38500, and that's related but NAB.
- [react-native-live-markdown is now working in the new arch](https://expensify.slack.com/archives/C04878MDF34/p1710783123960509)
- @j-piasecki removed a number of forks from the PR and moved them to patches per our best practices
- we're also dependent upon a RN upgrade again, because of react-native-live-markdown, but fortunately we've already been working on that for several weeks in https://github.com/Expensify/App/issues/37374

**Comment by tomekzaw** [comment]
> we're also dependent upon a RN upgrade again, because of react-native-live-markdown

We're able to backport the necessary changes to RN 0.73 to support Live Markdown on Fabric with RN 0.73.

At the moment, we need a patch to support Live Markdown on RN 0.74.

@j-piasecki can share more details on that

**Comment by roryabraham** [comment]
NEW ARCH PR HAS BEEN MERGED 🚀

**Comment by mallenexpensify** [comment]
Does @ishpaul777 need to be compensated for their PR review? If so, @roryabraham can you propose the amount and assign to me? I can take care of as BZ

**Comment by roryabraham** [comment]
yes, I'd propose a payment of $1000 for the review of this one, because it could so broadly affect any part of the app

**Comment by mallenexpensify** [comment]
I plan to pay @ishpaul777 $1000 7 days after the PR below hits production. 
- https://github.com/Expensify/App/pull/13767

**Comment by mallenexpensify** [comment]
I agree that $2000 seems more fair for the amount of time, testing and work involved.  We appreciate the help @ishpaul777 . 
25, 50 and 100% are mainly used for partial payments.   

For potential increases in price, i think it should be open-ended since requests are fairly rare and it allows the contributor, C+ and/or engineer to propose a price.   

PR hit staging an hour ago so we still have time here.

**Comment by Julesssss** [comment]
We have a couple of potential new issues here that might be related to the Fabric arch: 
- https://github.com/Expensify/App/issues/40077
- https://github.com/Expensify/App/issues/40048

**Comment by lakchote** [comment]
> We have a couple of potential new issues here that might be related to the Fabric arch:
> 
> * [[$500] iOS - Wallet - Bank button is not responsive #40077](https://github.com/Expensify/App/issues/40077)
> * [[$500] Android - LHN - Unable to open reports, tasks and threads in LHN #40048](https://github.com/Expensify/App/issues/40048)
> * [[$250] BA - Missing "Routing number" and "Account number" title after revisiting Bank info page #40082](https://github.com/Expensify/App/issues/40082)

There is also this one:

- https://github.com/Expensify/App/issues/40080

`react-native-view-shot `does not support new arch yet.
It'll only be fixed in the next version (0.74) of React Native (see https://github.com/gre/react-native-view-shot/pull/516).

**Comment by mallenexpensify** [comment]
@ishpaul777 , can you let me know when I should issue payment here? Lotta PRs keep auto-posting and pushing back the payment date.  Thx

**Comment by roryabraham** [comment]
I don't think it should, because it's an inconsistently reproducible bug in the RN Fabric renderer C++ code. I think we should pay this out on 2024-04-25, one week after the main Fabric PR went to prod (yesterday).

**Comment by mallenexpensify** [comment]
Groovy, I'll pay on 4/25, thx

**Comment by joekaufmanexpensify** [comment]
@mallenexpensify I see based on above comments you have been tracking payments/BZ for this one, so assigned to you. LMK if there is anything I can help with though!

**Comment by mallenexpensify** [comment]
@ishpaul777 can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~0112754e710d14595c

**Comment by mallenexpensify** [comment]
Contributor+: @ishpaul777 paid $2000 via Upwork. 

Thanks! Are there any new regression tests we want for this?  cc @roryabraham

**Comment by roryabraham** [comment]
No additional regression tests needed for this. Thanks for the great work everyone. This was a bit one. Closing this out!
